### ISSUE TITLE
xtest 9089 get failed when optee_client apply the patch   "Check flags before allocating shared memory" 

### ISSUE THREAD

**Issue by prime-zeng** [OP-TEE/optee_test#112]
```
* XTEST_TEE_9089 50-3c-14
DEBUG:   [0x0] TEE-CORE:tee_ta_init_static_ta_session:159:    Lookup for Static TA 534d4152-5443-534c-5443525950544f31
DEBUG:   [0x0] TEE-CORE:init_session_with_signed_ta:614:    Load dynamic TA
DEBUG:   [0x0] TEE-CORE:ta_load:396: Loaded TA at 0x7f100000
DEBUG:   [0x0] TEE-CORE:ta_load:397: ELF load address 0x101000
DEBUG:   [0x0] TEE-CORE:init_session_with_signed_ta:620:       dyn TA : 534d4152-5443-534c-5443525950544f31
FLOW:    [0x0] TEE-CORE:trace_syscall:144: syscall #15 (syscall_cryp_state_alloc)
FLOW:    [0x0] TEE-CORE:trace_syscall:144: syscall #18 (syscall_hash_init)
*** Error in `xtest': free(): invalid pointer: 0x0000007fefc071b0 ***
======= Backtrace: =========
/lib64/libc.so.6(+0x6b158)[0x7fb1585158]
/lib64/libc.so.6(+0x70e58)[0x7fb158ae58]
/lib64/libc.so.6(+0x71aa8)[0x7fb158baa8]
xtest[0x46f5a4]
======= Memory map: ========
00400000-00591000 r-xp 00000000 b3:07 115                                /bin/xtest
005a1000-005a5000 rw-p 00191000 b3:07 115                                /bin/xtest
005a5000-005a7000 rw-p 00000000 00:00 0 
04feb000-0500c000 rw-p 00000000 00:00 0                                  [heap]
7fac000000-7fac021000 rw-p 00000000 00:00 0 
7fac021000-7fb0000000 ---p 00000000 00:00 0 
7fb14e8000-7fb14f9000 r-xp 00000000 b3:07 264                            /lib64/libgcc_s.so.1
7fb14f9000-7fb1509000 ---p 00011000 b3:07 264                            /lib64/libgcc_s.so.1
7fb1509000-7fb150a000 rw-p 00011000 b3:07 264                            /lib64/libgcc_s.so.1
7fb150a000-7fb151a000 rw-p 00000000 00:00 0 
7fb151a000-7fb1665000 r-xp 00000000 b3:07 255                            /lib64/libc-2.22.so
7fb1665000-7fb1674000 ---p 0014b000 b3:07 255                            /lib64/libc-2.22.so
7fb1674000-7fb1678000 r--p 0014a000 b3:07 255                            /lib64/libc-2.22.so
7fb1678000-7fb167a000 rw-p 0014e000 b3:07 255                            /lib64/libc-2.22.so
7fb167a000-7fb167e000 rw-p 00000000 00:00 0 
7fb167e000-7fb1697000 r-xp 00000000 b3:07 285                            /lib64/libpthread-2.22.so
7fb1697000-7fb16a6000 ---p 00019000 b3:07 285                            /lib64/libpthread-2.22.so
7fb16a6000-7fb16a7000 r--p 00018000 b3:07 285                            /lib64/libpthread-2.22.so
7fb16a7000-7fb16a8000 rw-p 00019000 b3:07 285                            /lib64/libpthread-2.22.so
7fb16a8000-7fb16ac000 rw-p 00000000 00:00 0 
7fb16ac000-7fb16af000 r-xp 00000000 b3:07 624                            /lib64/libteec.so.1.0
7fb16af000-7fb16be000 ---p 00003000 b3:07 624                            /lib64/libteec.so.1.0
7fb16be000-7fb16bf000 rw-p 00002000 b3:07 624                            /lib64/libteec.so.1.0
7fb16bf000-7fb16de000 r-xp 00000000 b3:07 249                            /lib64/ld-2.22.so
7fb16e7000-7fb16eb000 rw-p 00000000 00:00 0 
7fb16eb000-7fb16ec000 r--p 00000000 00:00 0                              [vvar]
7fb16ec000-7fb16ed000 r-xp 00000000 00:00 0                              [vdso]
7fb16ed000-7fb16ee000 r--p 0001e000 b3:07 249                            /lib64/ld-2.22.so
7fb16ee000-7fb16f0000 rw-p 0001f000 b3:07 249                            /lib64/ld-2.22.so
7fefbe7000-7fefc08000 rw-p 00000000 00:00 0                              [stack]
DEBUG:   [0x0] TEE-CORE:tee_ta_close_session:318: tee_ta_close_session(0x7f04bf78)
DEBUG:   [0x0] TEE-CORE:tee_ta_close_session:337:    ... Destroy session
FLOW:    [0x0] TEE-CORE:trace_syscall:144: syscall #17 (syscall_cryp_state_free)
DEBUG:   [0x0] TEE-CORE:tee_ta_close_session:358:    ... Destroy TA ctx
Aborted
```

**Comment by prime-zeng** [comment]
This patch just do the simple check, but the xtest will get failed, maybe the bug of the xtest?

**Comment by jforissier** [comment]
Do you have optee_test patch: https://github.com/OP-TEE/optee_test/commit/5b41fbff33ff77b19836149fbde0a69524ced089?

`xtest 9089` is fine here on HiKey with the latest commits:
```
$ git submodule | grep optee_
+81f0257a4445338139b6ddc437a52f94c02c48fc optee_client (2.1.0-2-g81f0257)
+9fa2c2f1027c4e15c7bf4733522adba4d1c2fff9 optee_os (2.1.0-67-g9fa2c2f)
+b2865abeac9e9806a9ce785c4504ec1b7221fe50 optee_test (2.1.0-5-gb2865ab)
```

```
root@HiKey:/ xtest 9089
Test ID: 9089
Run test suite with level=0

TEE test application started with device [(null)]
######################################################
#
# XTEST_TEE_TEST
#
######################################################
 
* XTEST_TEE_9089 50-3c-14
DEBUG:   [0x0] TEE-CORE:tee_ta_init_static_ta_session:159:    Lookup for Static TA 534d4152-5443-534c-5443525950544f31
DEBUG:   [0x0] TEE-CORE:init_session_with_signed_ta:614:    Load dynamic TA
DEBUG:   [0x0] TEE-CORE:ta_load:396: Loaded TA at 0x3f200000
DEBUG:   [0x0] TEE-CORE:ta_load:397: ELF load address 0x40001000
DEBUG:   [0x0] TEE-CORE:init_session_with_signed_ta:620:       dyn TA : 534d4152-5443-534c-5443525950544f31
DEBUG:   [0x0] TEE-CORE:tee_ta_invoke_command:564:   => Error: ffff000a of 4
ERROR:   TEE-CORE: ERROR:   USER-TA: Panic 0x0
ERROR:   TEE-CORE: TA panicked with code 0x0 usr_sp 0x40000f38 usr_lr 0x40008277
DEBUG:   [0x0] TEE-CORE:user_ta_enter:525: tee_user_ta_enter: TA panicked with code 0x0
DEBUG:   [0x0] TEE-CORE:tee_ta_invoke_command:564:   => Error: ffff3024 of 3
DEBUG:   [0x0] TEE-CORE:tee_ta_invoke_command:540:    Panicked !
DEBUG:   [0x0] TEE-CORE:tee_ta_close_session:318: tee_ta_close_session(0x3f053f20)
DEBUG:   [0x0] TEE-CORE:tee_ta_close_session:337:    ... Destroy session
DEBUG:   [0x0] TEE-CORE:tee_ta_close_session:358:    ... Destroy TA ctx
  XTEST_TEE_9089 OK
+-----------------------------------------------------
Result of testsuite XTEST_TEE_TEST filtered by "9089":
XTEST_TEE_9089 OK
+-----------------------------------------------------
19 subtests of which 0 failed
1 test case of which 0 failed
723 test cases was skipped
TEE test application done!
root@HiKey:/ 
```

**Comment by prime-zeng** [comment]
Yes, I have the latest optee-test patched with the TEE_Initial_Configuration-Test_Suite_v1_1_0_4-2014_11_07,
And if I tried the optee_client with and with out 17d1add, 9089 will get abort with the patch and success without the patch.

**Comment by jforissier** [comment]
Something looks wrong on the non-secure side (xtest). Indeed, the error message is from libc:
```
*** Error in `xtest': free(): invalid pointer: 0x0000007fefc071b0 ***
```
It is the cause of the abort. Now you have to figure out the cause of the invalid `free()`. Perhaps the error condition from the TA panic is not handled properly? I will run `xtest 9089` with valgrind and double check the code and let you know if I find something.

**Comment by jforissier** [comment]
Yeah I think there's something wrong with the allocation/deallocation of `SHARE_MEM04`. It looks like we're trying things that have not been allocated. The code is hard to follow with all those macros and code generated from an `.xsl` (that's insane IMHO), so I did not investigate further. Here is the `valgrind` log FYI:

```
root@HiKey:/ valgrind xtest 9089
==905== Memcheck, a memory error detector
==905== Copyright (C) 2002-2015, and GNU GPL'd, by Julian Seward et al.
==905== Using Valgrind-3.12.0.SVN and LibVEX; rerun with -h for copyright info
==905== Command: xtest 9089
==905== 
Test ID: 9089
Run test suite with level=0

TEE test application started with device [(null)]
######################################################
#
# XTEST_TEE_TEST
#
######################################################
 
* XTEST_TEE_9089 50-3c-14
ERROR:   TEE-CORE: ERROR:   USER-TA: Panic 0x0
ERROR:   TEE-CORE: TA panicked with code 0x0 usr_sp 0x40000f38 usr_lr 0x40008277
==905== Conditional jump or move depends on uninitialised value(s)
==905==    at 0x485D4EC: TEEC_ReleaseSharedMemory (tee_client_api.c:678)
==905==    by 0x470C7B: mac_compute_final (xml_crypto_api.h:2586)
==905==    by 0x46EF0F: Invoke_Crypto_MACCompareFinal (xml_crypto_api.h:2026)
==905==    by 0x481A27: xtest_tee_9089 (xtest_9000.c:1643)
==905==    by 0x4244EF: ADBG_RunSuite (adbg_run.c:131)
==905==    by 0x4242D7: Do_ADBG_RunSuite (adbg_run.c:66)
==905==    by 0x421D83: main (xtest_main.c:183)
==905== 
==905== Conditional jump or move depends on uninitialised value(s)
==905==    at 0x485D500: TEEC_ReleaseSharedMemory (tee_client_api.c:680)
==905==    by 0x470C7B: mac_compute_final (xml_crypto_api.h:2586)
==905==    by 0x46EF0F: Invoke_Crypto_MACCompareFinal (xml_crypto_api.h:2026)
==905==    by 0x481A27: xtest_tee_9089 (xtest_9000.c:1643)
==905==    by 0x4244EF: ADBG_RunSuite (adbg_run.c:131)
==905==    by 0x4242D7: Do_ADBG_RunSuite (adbg_run.c:66)
==905==    by 0x421D83: main (xtest_main.c:183)
==905== 
==905== Syscall param munmap(start) contains uninitialised byte(s)
==905==    at 0x495D098: munmap (syscall-template.S:81)
==905==    by 0x485D533: TEEC_ReleaseSharedMemory (tee_client_api.c:684)
==905==    by 0x470C7B: mac_compute_final (xml_crypto_api.h:2586)
==905==    by 0x46EF0F: Invoke_Crypto_MACCompareFinal (xml_crypto_api.h:2026)
==905==    by 0x481A27: xtest_tee_9089 (xtest_9000.c:1643)
==905==    by 0x4244EF: ADBG_RunSuite (adbg_run.c:131)
==905==    by 0x4242D7: Do_ADBG_RunSuite (adbg_run.c:66)
==905==    by 0x421D83: main (xtest_main.c:183)
==905== 
==905== Syscall param munmap(length) contains uninitialised byte(s)
==905==    at 0x495D098: munmap (syscall-template.S:81)
==905==    by 0x485D533: TEEC_ReleaseSharedMemory (tee_client_api.c:684)
==905==    by 0x470C7B: mac_compute_final (xml_crypto_api.h:2586)
==905==    by 0x46EF0F: Invoke_Crypto_MACCompareFinal (xml_crypto_api.h:2026)
==905==    by 0x481A27: xtest_tee_9089 (xtest_9000.c:1643)
==905==    by 0x4244EF: ADBG_RunSuite (adbg_run.c:131)
==905==    by 0x4242D7: Do_ADBG_RunSuite (adbg_run.c:66)
==905==    by 0x421D83: main (xtest_main.c:183)
==905== 
==905== Conditional jump or move depends on uninitialised value(s)
==905==    at 0x485D4EC: TEEC_ReleaseSharedMemory (tee_client_api.c:678)
==905==    by 0x470C8B: mac_compute_final (xml_crypto_api.h:2587)
==905==    by 0x46EF0F: Invoke_Crypto_MACCompareFinal (xml_crypto_api.h:2026)
==905==    by 0x481A27: xtest_tee_9089 (xtest_9000.c:1643)
==905==    by 0x4244EF: ADBG_RunSuite (adbg_run.c:131)
==905==    by 0x4242D7: Do_ADBG_RunSuite (adbg_run.c:66)
==905==    by 0x421D83: main (xtest_main.c:183)
==905== 
==905== Conditional jump or move depends on uninitialised value(s)
==905==    at 0x485D500: TEEC_ReleaseSharedMemory (tee_client_api.c:680)
==905==    by 0x470C8B: mac_compute_final (xml_crypto_api.h:2587)
==905==    by 0x46EF0F: Invoke_Crypto_MACCompareFinal (xml_crypto_api.h:2026)
==905==    by 0x481A27: xtest_tee_9089 (xtest_9000.c:1643)
==905==    by 0x4244EF: ADBG_RunSuite (adbg_run.c:131)
==905==    by 0x4242D7: Do_ADBG_RunSuite (adbg_run.c:66)
==905==    by 0x421D83: main (xtest_main.c:183)
==905== 
==905== Syscall param munmap(start) contains uninitialised byte(s)
==905==    at 0x495D098: munmap (syscall-template.S:81)
==905==    by 0x485D533: TEEC_ReleaseSharedMemory (tee_client_api.c:684)
==905==    by 0x470C8B: mac_compute_final (xml_crypto_api.h:2587)
==905==    by 0x46EF0F: Invoke_Crypto_MACCompareFinal (xml_crypto_api.h:2026)
==905==    by 0x481A27: xtest_tee_9089 (xtest_9000.c:1643)
==905==    by 0x4244EF: ADBG_RunSuite (adbg_run.c:131)
==905==    by 0x4242D7: Do_ADBG_RunSuite (adbg_run.c:66)
==905==    by 0x421D83: main (xtest_main.c:183)
==905== 
==905== Syscall param munmap(length) contains uninitialised byte(s)
==905==    at 0x495D098: munmap (syscall-template.S:81)
==905==    by 0x485D533: TEEC_ReleaseSharedMemory (tee_client_api.c:684)
==905==    by 0x470C8B: mac_compute_final (xml_crypto_api.h:2587)
==905==    by 0x46EF0F: Invoke_Crypto_MACCompareFinal (xml_crypto_api.h:2026)
==905==    by 0x481A27: xtest_tee_9089 (xtest_9000.c:1643)
==905==    by 0x4244EF: ADBG_RunSuite (adbg_run.c:131)
==905==    by 0x4242D7: Do_ADBG_RunSuite (adbg_run.c:66)
==905==    by 0x421D83: main (xtest_main.c:183)
==905== 
  XTEST_TEE_9089 OK
+-----------------------------------------------------
Result of testsuite XTEST_TEE_TEST filtered by "9089":
XTEST_TEE_9089 OK
+-----------------------------------------------------
19 subtests of which 0 failed
1 test case of which 0 failed
723 test cases was skipped
TEE test application done!
==905== 
==905== HEAP SUMMARY:
==905==     in use at exit: 308 bytes in 14 blocks
==905==   total heap usage: 19 allocs, 5 frees, 636 bytes allocated
==905== 
==905== LEAK SUMMARY:
==905==    definitely lost: 0 bytes in 0 blocks
==905==    indirectly lost: 0 bytes in 0 blocks
==905==      possibly lost: 0 bytes in 0 blocks
==905==    still reachable: 308 bytes in 14 blocks
==905==         suppressed: 0 bytes in 0 blocks
==905== Rerun with --leak-check=full to see details of leaked memory
==905== 
==905== For counts of detected and suppressed errors, rerun with: -v
==905== Use --track-origins=yes to see where uninitialised values come from
==905== ERROR
```

**Comment by prime-zeng** [comment]
So there is indeed something wrong with the the memory operation? i think i need to dig into it when i am free.

**Comment by jforissier** [comment]
@prime-zeng yes, a bug in xtest, that's what I think.

**Comment by prime-zeng** [comment]
Some of the reason is that TEEC_ReleaseSharedMemory don't do enough check for the input parameters.
when the TEEC_AllocateSharedMemory get failed, the result is undefined.

**Comment by prime-zeng** [comment]
I find the reason of the abort ：
In function Invoke_Crypto_MACCompareFinal

when the ALLOCATE_SHARED_MEMORY, it will jump to exit:
` exit:
     CRYPTO_FREE(mac);
     TEEC_ReleaseSharedMemory(SHARE_MEM01);
     TEEC_ReleaseSharedMemory(SHARE_MEM02);
`
CRYPTO_FREE(mac) will result an abort. 

But i don't know how to fix it, since the code is not part of the optee_test.

**Comment by jforissier** [comment]
What whould you change in `TEEC_ReleaseSharedMemory()`? In my opinion, it's fine. `TEEC_AllocateSharedMemory()` returns an error status when something fails, then you're not supposed to call `TEEC_ReleaseSharedMemory()` on the (invalid) TEEC_SharedMemory pointer.

About `CRYPTO_FREE(mac);` in `Invoke_Crypto_MACCompareFinal()`: you're right, `CRYPTO_INIT(mac)` should be done sooner (before the shared memory allocations). What prevents you from changing it? `host/xtest/xml/include/xml_crypto_api.h` is part of `optee_test`.

**Comment by prime-zeng** [comment]
in the xtest,when the TEEC_AllocateSharedMemory() get failed, it will directly go to the TEEC_ReleaseSharedMemory(), so TEEC_ReleaseSharedMemory() will need to process uninitialized input.

I got it wrong, i  thought Invoke_Crypto_MACCompareFinal was in the xtest_9000.c, in fact , it is in the 
xml_crypto_api.h, i will fix it.

**Comment by jforissier** [comment]
On Fri, Aug 19, 2016 at 10:46 AM, prime-zeng <notifications@github.com>
wrote:

> in the xtest,when the TEEC_AllocateSharedMemory() get failed, it will
> directly go to the TEEC_ReleaseSharedMemory(), so
> TEEC_ReleaseSharedMemory() will need to process uninitialized input.
>
>
Solution: don't call TEEC_ReleaseSharedMemory() when
TEEC_AllocateSharedMemory() fails, obviously. xtest needs fixing.

**Comment by prime-zeng** [comment]
agree, but i think there is to much fix for error handle. And i still think TEEC_ReleaseSharedMemory should have more check.

**Comment by igoropaniuk** [comment]
Hi @prime-zeng, could you please check if this issue is still reproduced with the latest changes [5b3faa0](https://github.com/OP-TEE/optee_test/commit/5b3faa07c4d522ee2a2358f42c794a09c55afb56) 

Thanks

**Comment by jforissier** [comment]
@igoropaniuk FYI, the valgrind errors I reported above are now gone.

**Comment by igoropaniuk** [comment]
@jforissier thanks for testing!

**Comment by etienne-lms** [comment]
Fix ref: https://github.com/OP-TEE/optee_test/pull/133

**Comment by igoropaniuk** [comment]
Hi @prime-zeng , still waiting for your response

**Comment by prime-zeng** [comment]
@igoropaniuk , ok, i will test it later this week.

**Comment by igoropaniuk** [comment]
@prime-zeng thanks!

**Comment by prime-zeng** [comment]
@igoropaniuk  i just down load the latest code, and it runs ok on my platform, do i need extra confirm?

**Comment by igoropaniuk** [comment]
@prime-zeng thanks for testing! ;)
